### ISSUE TITLE
Collections fail to decode consistently

### ISSUE THREAD

**Issue by achubbic-snav** [gabhijit/hampi#70]
Hello, I have encountered an issue in which messages containing collections (vectors in this case) of submessages fail to decode properly when the number of elements in the collection changes (for test purposes, I have experimented with creating a collection of identical elements, to ensure that it is not something unique to a given element that causes failure).

I have tried to provide as much context as possible, but it is difficult to produce a minimal reproducible example in this case. I will comment more on this at the end of this post.

I have a data type which abides by the form:
```rust
#[derive(asn1_codecs_derive :: UperCodec, Debug)]
#[asn(type = "SEQUENCE", extensible = true)]
pub struct GNSS_SSR_OrbitCorrections_r15 {
    pub epoch_time_r15: GNSS_SystemTime,
    pub ssr_update_interval_r15: INTEGER_580,
    pub satellite_reference_datum_r15: ENUMERATED_581,
    pub iod_ssr_r15: INTEGER_582,
    pub ssr_orbit_correction_list_r15: SSR_OrbitCorrectionList_r15, // <- where I believe the problem lies
}
```

I do not believe that many of these fields are relevant, but know that `SSR_OrbitCorrectionList_r15` is a collection which takes the form:
```rust
#[derive(asn1_codecs_derive :: UperCodec, Debug)]
#[asn(type = "SEQUENCE-OF", sz_extensible = false, sz_lb = "1", sz_ub = "64")]
pub struct SSR_OrbitCorrectionList_r15(pub Vec<SSR_OrbitCorrectionSatelliteElement_r15>);
```
to ensure consistentcy, elements of `SSR_OrbitCorrectionList_r15` (type `SSR_OrbitCorrectionSatelliteElement_r15`) are duplicates of one another. ```SSR_OrbitCorrectionSatelliteElement_r15``` contains primitive data types.

Upon encountering a decode failure, I wrote a test to verify that encoding and subsequent decoding of a message of a collection type is successful

```rust
// verify that encoding and subsequent decoding of a message of a collection type is successful
pub fn encode_decode_sanity_check_orbit_correction_list(
    orbit_corrections: GNSS_SSR_OrbitCorrections_r15,
) -> Result<()> {
    let mut data: PerCodecData = PerCodecData::new_uper();

    orbit_corrections
        .ssr_orbit_correction_list_r15
        .uper_encode(&mut data)?;

    let bytes = data.into_bytes();
    println!("bytes.len(): {}", bytes.len());

    let mut decode_data: PerCodecData = PerCodecData::from_slice_uper(&bytes);

    let msg = SSR_OrbitCorrectionList_r15::uper_decode(&mut decode_data)?;

    Ok(())
}
```

This function SOMETIMES returns Ok(()). Most of the time it returns 
`Error: PerCodec:GetBitError:Requested Bit <something>, Remaining bits <something else>`

By modifying ONLY the number of elements in ` SSR_OrbitCorrectionList_r15,` (`Vec<SSR_OrbitCorrectionSatelliteElement_r15>`), I can achieve different results:
- sometimes the decoding fails with the error described above
- sometimes the decoding succeeds but the data has changed from what I encoded
- sometimes the encoding succeeds nominally, though this is rare

I suspect that this problem relates to collections because I have performed this "roundtrip" process on all other subfields of the parent `GNSS_SSR_OrbitCorrections_r15` structure and have not encountered this problem.

An additional piece of information you may find useful is that if I attempted to encode then immediately decode each element of the vector, each of these roundtrips would succeed. It is only when I try to encode the vector `SSR_OrbitCorrectionList_r15`, or the vector's parent `GNSS_SSR_OrbitCorrections_r15` that failure occurs.

Please let me know how I can be of assistance. If the information I have provided is insufficient I would be happy to send you with further information related to the bindings I am using, tests, etc

**Comment by achubbic-snav** [comment]
@gabhijit attached please find a minimal reproducible example, along with a bindings file describing message structure.
[inconsistent_hampi_decoding.zip](https://github.com/gabhijit/hampi/files/10396432/inconsistent_hampi_decoding.zip)

In its present state, `cargo run` will perform a simple encode/decode process successfully. You can prompt a failure by changing the upper range value in `main.rs`'s `for` loop on line 67.

The requested hex bytes are printed each run

**Comment by gabhijit** [comment]
@achubbic-snav : Thanks this indeed seems to be a bug somewhere in the `uper` codec. Will take a look. For lengths more than 2, this consistently fails.

**Comment by gabhijit** [comment]
@achubbic-snav : There was an issue with the codec that didn't do it right for `constrained_whole_numbers`. Pushed a fix for this to master. You might want to try with latest master branch. Let me know if this works fine for you.

**Comment by achubbic-snav** [comment]
Upon doing so my testing as so far been successful. Thank you.

**Comment by gabhijit** [comment]
Great! Will publish this change in the next crate release (0.5.6 in a couple of days). Till such time, you might have to work with git based dependency! Hope that works.

**Comment by gabhijit** [comment]
Closing this issue as the fix #73 is merged into master.
